server/handlers/validation_test.go

```diff
@@ -0,0 +1,893 @@
+	_, repo, _ := testutils.EmptyRepo()
+	store := storage.NewMemStorage()
+
+	r, tg, sn, ts, err := testutils.Sign(repo)
+	assert.NoError(t, err)
+
+	// add another copy of the signature so the hash is different
+	tg.Signatures = append(tg.Signatures, tg.Signatures[0])
+
+	root, targets, snapshot, timestamp, err := testutils.Serialize(r, tg, sn, ts)
+	assert.NoError(t, err)
+
+	updates := []storage.MetaUpdate{
+		{
+			Role:    "root",
+			Version: 1,
+			Data:    root,
+		},
+		{
+			Role:    "targets",
+			Version: 1,
+			Data:    targets,
+		},
+		{
+			Role:    "snapshot",
+			Version: 1,
+			Data:    snapshot,
+		},
+		{
+			Role:    "timestamp",
+			Version: 1,
+			Data:    timestamp,
+		},
+	}
+
+	err = validateUpdate("testGUN", updates, store)
+	assert.Error(t, err)
+	assert.IsType(t, ErrBadSnapshot{}, err)
+}
+
+// ### End snapshot size mismatch negative tests ###
+
+// ### Snapshot hash mismatch negative tests ###
+func TestValidateRootModifiedHash(t *testing.T) {
+	_, repo, _ := testutils.EmptyRepo()
+	store := storage.NewMemStorage()
+
+	r, tg, sn, ts, err := testutils.Sign(repo)
+	assert.NoError(t, err)
+
+	snap, err := data.SnapshotFromSigned(sn)
+	assert.NoError(t, err)
+	snap.Signed.Meta["root"].Hashes["sha256"][0] = snap.Signed.Meta["root"].Hashes["sha256"][0] ^ 0xff
+
+	sn, err = snap.ToSigned()
+	assert.NoError(t, err)
+
+	root, targets, snapshot, timestamp, err := testutils.Serialize(r, tg, sn, ts)
+	assert.NoError(t, err)
+
+	updates := []storage.MetaUpdate{
+		{
+			Role:    "root",
+			Version: 1,
+			Data:    root,
+		},
+		{
+			Role:    "targets",
+			Version: 1,
+			Data:    targets,
+		},
+		{
+			Role:    "snapshot",
+			Version: 1,
+			Data:    snapshot,
+		},
+		{
+			Role:    "timestamp",
+			Version: 1,
+			Data:    timestamp,
+		},
+	}
+
+	err = validateUpdate("testGUN", updates, store)
+	assert.Error(t, err)
+	assert.IsType(t, ErrBadSnapshot{}, err)
+}
+
+func TestValidateTargetsModifiedHash(t *testing.T) {
+	_, repo, _ := testutils.EmptyRepo()
+	store := storage.NewMemStorage()
+
+	r, tg, sn, ts, err := testutils.Sign(repo)
+	assert.NoError(t, err)
+
+	snap, err := data.SnapshotFromSigned(sn)
+	assert.NoError(t, err)
+	snap.Signed.Meta["targets"].Hashes["sha256"][0] = snap.Signed.Meta["targets"].Hashes["sha256"][0] ^ 0xff
+
+	sn, err = snap.ToSigned()
+	assert.NoError(t, err)
+
+	root, targets, snapshot, timestamp, err := testutils.Serialize(r, tg, sn, ts)
+	assert.NoError(t, err)
+
+	updates := []storage.MetaUpdate{
+		{
+			Role:    "root",
+			Version: 1,
+			Data:    root,
+		},
+		{
+			Role:    "targets",
+			Version: 1,
+			Data:    targets,
+		},
+		{
+			Role:    "snapshot",
+			Version: 1,
+			Data:    snapshot,
+		},
+		{
+			Role:    "timestamp",
+			Version: 1,
+			Data:    timestamp,
+		},
+	}
+
+	err = validateUpdate("testGUN", updates, store)
+	assert.Error(t, err)
+	assert.IsType(t, ErrBadSnapshot{}, err)
+}
+
+// ### End snapshot hash mismatch negative tests ###
```

Comment:
The test cases look really extensive. Out of curiousity, what level of test coverage are you seeing?